linera-execution/src/evm/database.rs

```diff
@@ -115,21 +107,17 @@ fn application_id_to_address(application_id: ApplicationId) -> Address {
 impl<Runtime: BaseRuntime> DatabaseRuntime<Runtime> {
     /// Encode the `index` of the EVM storage associated to the smart contract
     /// in a linera key.
-    fn get_linera_key(val: u8, index: U256) -> Result<Vec<u8>, ExecutionError> {
-        let mut key = vec![val, KeyCategory::Storage as u8];
+    fn get_linera_key(key_prefix: &[u8], index: U256) -> Result<Vec<u8>, ExecutionError> {
+        let mut key = key_prefix.to_vec();
         bcs::serialize_into(&mut key, &index)?;
         Ok(key)
```

Comment:
How can you be sure this is safe? `serialize_into` will write into `mut key` and potentially overwrite content with `&index` value.

Reply:
The function `serialize_into` will append the content of the serialization into `key`. So, I am not sure where any overwriting will occur.

Reply:
It depends on the implementation of `std::io::Write` for this type
```rust
/// Same as `to_bytes` but write directly into an `std::io::Write` object.
pub fn serialize_into<W, T>(write: &mut W, value: &T) -> Result<()>
where
    W: ?Sized + std::io::Write,
    T: ?Sized + Serialize,
{
    let serializer = Serializer::new(write, crate::MAX_CONTAINER_DEPTH);
    value.serialize(serializer)
}
```

Reply:
I am not sure this is typical though. The first argument is a target for writing a serialized value. I will try to find an example tomorrow.

Reply:
Intuitively:
* The signature of this method from `bcs` could and should probably be `serialize_into<W, T>(write: W, ...)` then it would be clear that we can only append to the `write` object.
* The blanket implementation of `Write` for `&mut W` is most certainly just forwarding the writes.